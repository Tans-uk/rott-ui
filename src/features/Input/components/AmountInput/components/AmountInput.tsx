import {useCallback, useEffect, useRef, useState, type FC} from 'react'

import {StyleSheet, TextInput} from 'react-native'

import type {AmountInputProps} from '../models'
import {AmountInputStyles} from '../styles'

import {Icon, type IconTypes} from '@features/Icon'
import {InputStyleNormalizer, InputStyles} from '@features/Input'
import {Item} from '@features/Item'
import {Label} from '@features/Label'
import {themeConfig} from '@providers'

export const AmountInput: FC<AmountInputProps> = ({
  fontSize,
  onChangeText,
  value,
  theme,
  disabled,
  size,
  currencyType = 'TL',
  ...props
}) => {
  const amountRef = useRef<TextInput>(null)
  const currencyRef = useRef<TextInput>(null)
  const [amount, setAmount] = useState('')
  const [currency, setCurrency] = useState('')

  const resetInternalStates = useCallback(() => {
    setAmount('')
    setCurrency('')
    onChangeText!('0.00')
  }, [])

  const replaceTextWithNumberOrEmpty = (text: string) => {
    const numericOnly = text.replace(/[^0-9]/g, '')

    if (numericOnly === '0') return numericOnly

    return numericOnly.replace(/^0+/, '')
  }

  const amountNormalizer = (text: string) => {
    if (!text) return resetInternalStates()

    const replacedAmount = replaceTextWithNumberOrEmpty(text)
    const reversedText = replacedAmount.split('').reverse().join('') ?? ''
    const dottedText = reversedText?.match(/.{1,3}/g)?.join('.')
    const normalizedText = dottedText?.split('').reverse().join('')

    setAmount(normalizedText ?? '')
  }

  const formatCurrency = (text: string) => {
    return text.replace(/[^0-9]/g, '')
  }

  const handleTextChange = (inputTextAmount: string, inputTextCurrency: string) => {
    if (!inputTextAmount && !inputTextCurrency) return

    let amountFormat = replaceTextWithNumberOrEmpty(inputTextAmount)
    let currencyFormat = formatCurrency(inputTextCurrency)

    amountFormat = amountFormat.length === 0 ? '0' : amountFormat

    currencyFormat = currencyFormat.length === 0 ? '00' : currencyFormat
    currencyFormat = currencyFormat.length === 1 ? `${currencyFormat}0` : currencyFormat

    if (currencyFormat.length > 2) currencyFormat = currencyFormat.substring(0, 2)

    onChangeText!(`${amountFormat}.${currencyFormat}`)
  }

  const placeholderColorNormalizer =
    theme === 'dark'
      ? themeConfig.colors.white
      : (!amount && !currency) ||
          ((amount === '0' || amount === '') && (!currency || currency === '' || currency === '00'))
        ? themeConfig.colors['grey-200']
        : themeConfig.colors['grey-900']

  useEffect(() => {
    const initialAmount = value?.split(',')[0]
    const initialCurrency = value?.split(',')[1]

    setAmount(initialAmount ?? '')
    setCurrency(initialCurrency ?? '')
  }, [])

  useEffect(() => {
    if (value === '' || !value || value === '0.00' || value === '0') resetInternalStates()
  }, [value])

  useEffect(() => {
    handleTextChange(amount, currency)
  }, [amount, currency])

  return (
    <Item relative onTouchStart={() => amountRef.current?.focus()}>
      <Item row {...props}>
        <TextInput
          ref={amountRef}
          nativeID='amount-native-id'
          testID='amount-test-id'
          editable={!disabled}
          placeholder='0'
          maxLength={11}
          value={amount}
          keyboardType='number-pad'
          onChangeText={(text) => amountNormalizer(text)}
          style={StyleSheet.flatten([
            InputStyles({fontSize, theme, size}).defaultTextInputStyle,
            AmountInputStyles().amountInputStyle,
            {
              color: placeholderColorNormalizer,
            },
          ])}
          placeholderTextColor={placeholderColorNormalizer}
        />

        <Item
          justifyContentFlexEnd
          paddingBottom={InputStyleNormalizer({size}).bottomElementPadding}>
          <Label
            fontSize={
              fontSize ??
              InputStyleNormalizer({
                size,
              }).placeholderSize
            }
            fontWeight='bold'
            color={placeholderColorNormalizer as string}>
            ,
          </Label>
        </Item>

        <TextInput
          ref={currencyRef}
          testID='currency-test-id'
          maxLength={2}
          placeholder='00'
          value={currency}
          keyboardType='number-pad'
          onChangeText={(text) => {
            if (text.isEmpty()) amountRef.current?.focus()
            setCurrency(formatCurrency(text).substring(0, 2))
          }}
          onKeyPress={({nativeEvent}) => {
            if (nativeEvent.key === 'Backspace' && currency.isEmpty()) amountRef.current?.focus()
          }}
          onFocus={() => (currency === '00' ? setCurrency('') : undefined)}
          onBlur={() => (currency.isEmpty() || currency === '0' ? setCurrency('00') : undefined)}
          style={StyleSheet.flatten([
            InputStyles({fontSize, theme, size}).defaultTextInputStyle,
            AmountInputStyles().amountInputStyle,
            {
              color: placeholderColorNormalizer,
            },
          ])}
          placeholderTextColor={placeholderColorNormalizer}
        />
      </Item>

      <Item absolute right={0} bottom={InputStyleNormalizer({size}).icon.paddingBottom}>
        <Icon
          name={currencyType as IconTypes}
          width={InputStyleNormalizer({size}).icon.width}
          height={InputStyleNormalizer({size}).icon.height}
          color={themeConfig.colors['grey-200']}
          mode='fill'
          noStroke
        />
      </Item>
    </Item>
  )
}
