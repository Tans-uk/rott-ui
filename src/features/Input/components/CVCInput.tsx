import {type FC} from 'react'

import {StyleSheet, TextInput} from 'react-native'

import {formatMessage, translator} from '../../../libs'
import {ThemeConfig} from '../../../models'
import {AlertDialog} from '../../AlertDialog'
import {Icon} from '../../Icon'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Pressable} from '../../Pressable'
import type {CVCInputProps} from '../models'
import {useCVCInputStyles, useInputStyles} from '../styles'

export const CVCInput: FC<CVCInputProps<ThemeConfig>> = ({
  fontSize,
  onChangeText,
  theme,
  size,
  ...props
}) => {
  const {defaultTextInputStyle} = useInputStyles({fontSize, theme, size})
  const {infoIcon} = useCVCInputStyles()
  const handleTextChange = (inputText: string) => {
    let text = inputText.replace(/[^0-9]/g, '')
    text = text.length > 3 ? text.substring(0, 3) : text
    onChangeText!(text)
  }

  return (
    <Item row>
      <TextInput
        editable={!props.disabled}
        placeholder='***'
        style={StyleSheet.flatten([defaultTextInputStyle])}
        keyboardType='number-pad'
        maxLength={3}
        onChangeText={handleTextChange}
        {...props}
      />

      {/* TODO: CVC Info butonuna kart arkayüzü eklenmeli */}
      <Pressable
        testID='info-icon-test-id'
        style={infoIcon}
        row
        justifyContentCenter
        alignItemsCenter
        onPress={() => {
          AlertDialog.show({
            title: formatMessage('CVC.INFO.TITLE'),
            text: formatMessage('CVC.INFO.DESCRIPTION'),
            buttons: {
              cancelButton: {
                text: 'COMMON.OK',
                variant: 'primary',
                onPress: () => AlertDialog.hide(),
              },
            },
          })
        }}>
        <Label variant='grey-200' fontSize='xs' fontWeight={'700'}>
          {translator('CVC.LABEL.DESCRIPTION')}
        </Label>
        <Icon name='CHECK_CIRCLE' height={18} width={18} variant='grey-200' marginLeft={4} />
      </Pressable>
    </Item>
  )
}
