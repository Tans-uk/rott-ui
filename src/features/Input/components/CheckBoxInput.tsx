/* eslint-disable react-native/no-inline-styles */
import {type FC} from 'react'

import {StyleSheet} from 'react-native'

import {themeConfig} from '../../../providers'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Pressable} from '../../Pressable'
import {type CheckBoxInputProps} from '../models'
import {InputStyles} from '../styles'
import {InputStyleNormalizer} from '../utils'
import React from 'react'

export const CheckBoxInput: FC<CheckBoxInputProps> = ({
  description,
  checked,
  fontSize,
  theme,
  size,
  disabled,
  variant,
  fontWeight,
  onCheckChange,
  testID,
  ...props
}) => {
  return (
    <Pressable
      flex={0}
      testID={testID ?? 'checkbox-input-test-id'}
      onPress={() => !disabled && onCheckChange && onCheckChange(!checked)}
      {...props}>
      <Item
        row
        alignItemsCenter
        style={[
          StyleSheet.flatten([InputStyles({theme, size}).defaultTextInputStyle, {height: 'auto'}]),
        ]}
        testID='checkbox-container-test-id'>
        <Item
          borderRadius={8}
          borderColor={checked ? themeConfig.colors.primary : themeConfig.colors['grey-200']}
          borderWidth={2}
          width={24}
          height={24}
          alignItemsCenter
          justifyContentCenter>
          {(checked || disabled) && (
            <Item
              testID='checkbox-checked-test-id'
              borderWidth={2}
              borderColor={
                checked ? themeConfig.colors['neutral-blue-soft'] : themeConfig.colors['grey-200']
              }
              borderRadius={4}
              backgroundColor={
                checked ? themeConfig.colors.primary : themeConfig.colors['grey-200']
              }
              width={checked ? 15 : 22}
              height={checked ? 15 : 22}
            />
          )}
        </Item>

        {description && typeof description === 'string' && (
          <Item flex={1} style={{flexDirection: 'row'}}>
            <Label
              testID='checkbox-default-label-test-id'
              fontSize={fontSize ?? InputStyleNormalizer({size}).placeholderSize}
              fontWeight={fontWeight ?? 'normal'}
              marginLeft={12}
              variant={variant ?? (theme === 'dark' ? 'white' : 'grey-900')}
              style={{flexWrap: 'wrap'}}>
              {description}
            </Label>
          </Item>
        )}

        {description && typeof description !== 'string' && <>{description}</>}
      </Item>
    </Pressable>
  )
}
