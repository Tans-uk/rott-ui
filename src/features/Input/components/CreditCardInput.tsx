import React, {type FC} from 'react'

import {StyleSheet} from 'react-native'

import type {CreditCardInputProps} from '../models'
import {InputStyles} from '../styles'
import {InputContainer} from './InputContainer'

// Package Imports
import MaskInput from 'react-native-mask-input'

export const CreditCardInput: FC<CreditCardInputProps> = ({
  fontSize,
  onChangeText,
  theme,
  disabled,
  size,
  ...props
}) => {
  const MASK = [
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
    '-',
    /\d/,
    /\d/,
    /\d/,
    /\d/,
  ]

  const handleOnChangeText = (text: string) => {
    const replacedText = text.replace(/[^0-9]/g, '')

    onChangeText!(replacedText)
  }

  return (
    <InputContainer {...props} size={size} theme={theme}>
      <MaskInput
        editable={!disabled}
        mask={MASK}
        placeholder='**** **** **** ****'
        keyboardType='number-pad'
        maxLength={19}
        onChangeText={(_masked, unmasked) => handleOnChangeText(unmasked)}
        style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
        {...props}
      />
    </InputContainer>
  )
}
