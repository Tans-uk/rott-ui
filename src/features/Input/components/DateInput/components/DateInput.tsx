// React Imports
import {useContext, useEffect, useRef, type FC} from 'react'

import {StyleSheet} from 'react-native'

import {RottUiContext} from '../../../../../contexts'
import {useDisplay} from '../../../../../hooks'
import {formatMessage} from '../../../../../libs'
import {ModalIdEnum, ThemeConfig} from '../../../../../models'
import {Button} from '../../../../Button'
import {Icon} from '../../../../Icon'
import {Item} from '../../../../Item'
import {Label} from '../../../../Label'
import {List} from '../../../../List'
import {Modal, useModal} from '../../../../Modal'
import {Pressable} from '../../../../Pressable'
import {Separator} from '../../../../Separator'
import {DateInputStyles, useInputStyles} from '../../../styles'
import {formatByDateMode, useInputStyleNormalizer} from '../../../utils'
import type {DataModel, DateInputProps} from '../models'

import {startOfDay} from 'date-fns'

import DatePicker from 'react-native-date-picker'

/**
 *
 * DateInput'un kullanımı diğer inputlara göre biraz daha farklı.
 * Yapısı gereği diğer inputlarda kullandığımız "onChangeText" event'i ile değişen değeri almamız mümkün olmadığı için
 * formik'in içinde yer alan "setFieldValue" metodunu kullanmamız gerek.
 *
 * ### **Örnek**:
 *   <Input
      label='Date Deneme'
      name='order-date'
      onDateChange={(date) => setFieldValue('orderDate', date)}
      type='date'
      mode='modal-date'
      value={orderDate}
      data={[
        {
          label: 'Bugün',
          value: new Date(),
        },
        {
          label: 'Yarın',
          value: add(new Date(), {days: 1}),
        },
      ]}
    />
 *
 * @property {(date: Date) => void)} onDateChange: seçilen tarih değerini almak içim bu property'nin kullanılması gerekmektedir.
 * @property {'date' | 'time' | 'datetime'} mode: İstenilen mod neyse ona göre belirtilmelidir. Default değer datetime olarak ayarlanmıştır.
 * @property {boolean} dontAllowClearing: Eğer date input'un değerinin temizlenmesine izin verilmemesi gerekiyorsa bu propert kullanılmalıdır.
 * dontAllowClearing={true} yerine sadece downAllowClearing olarak verirseniz değer otomatik olarak true kabul edilir. False olmadığı durumlarda
 * lütfen ={true} şeklinde tanımlama yapmayınız.
 */
let externalDate = new Date()
export const DateInput: FC<DateInputProps<ThemeConfig>> = ({
  placeholder = 'Lütfen bir tarih seçiniz',
  onDateChange,
  mode = 'date',
  allowClear = false,
  minimumDate,
  maximumDate,
  value,
  testID,
  theme,
  data,
  disabled,
  size,
  viewType = 'input',
  ...props
}) => {
  const {colors} = useContext(RottUiContext)
  const selectedItem = useRef<DataModel>(undefined)
  const {defaultTextInputStyle} = useInputStyles({theme, size, includeBorderRadius: true, ...props})
  const {setHeight} = useDisplay()
  const handleConfirmPress = (date?: Date | DataModel) => {
    let validDate: Date

    const isDataModel = date && 'label' in date
    const selectedDate = isDataModel ? date.value : date

    if (
      minimumDate &&
      (selectedDate ? selectedDate.getTime() : externalDate.getTime()) < minimumDate?.getTime()
    )
      validDate = minimumDate
    else if (
      maximumDate &&
      (selectedDate ? selectedDate.getTime() : externalDate.getTime()) > maximumDate?.getTime()
    )
      validDate = maximumDate
    else validDate = selectedDate ?? externalDate

    validDate = mode.includes('time') ? validDate : startOfDay(validDate)

    onDateChange!(validDate)
    selectedItem.current = isDataModel ? {...date, value: validDate} : {label: '', value: validDate}
    Modal.hideModal(ModalIdEnum.NativeDatePicker)
  }

  const handleClearPress = () => {
    Modal.hideModal()
    externalDate = new Date()
    onDateChange!(null as any)
  }

  const isSameDay = (date1: Date, date2: Date) =>
    date1?.getDate() === date2?.getDate() &&
    date1?.getMonth() === date2?.getMonth() &&
    date1?.getFullYear() === date2?.getFullYear()

  const isSelected = (item: DataModel) => {
    if (!item.value || !value) return
    const valueAsDate = new Date(value)

    if (!selectedItem.current) return isSameDay(item.value, valueAsDate)

    return isSameDay(item.value, valueAsDate) && selectedItem.current.label === item.label
  }

  const modalHeightCalculator = (dataCount?: number) => {
    if (dataCount === 0 || !dataCount) return 30

    const headerHeightInPercentage = 12.1 // 102 / 8.44
    const dataHeight = (dataCount + 2) * 52 // +1 for the default date picker item
    const calculatedModalHeightPercentage = dataHeight / 8.44 + headerHeightInPercentage

    return calculatedModalHeightPercentage >= 100 ? 100 : calculatedModalHeightPercentage
  }
  const {showModal: showNativeDatePicker, hideModal} = useModal(
    {
      id: ModalIdEnum.NativeDatePicker,
      testID: 'date-input-modal',
      visible: true,
      height: modalHeightCalculator(),
      onClose: () => {
        hideModal(ModalIdEnum.NativeDatePicker)
      },
      disableOutsideClick: false,
      backgroundColor: colors.white,
      header: (
        <Item
          justifyContentCenter
          backgroundColor={colors['grey-900']}
          height={50}
          style={DateInputStyles().dateInputHeaderStyle}>
          <Pressable
            testID='date-input-confirm-button'
            onPress={() => handleConfirmPress()}
            text={formatMessage('COMMON.OK')}
            textVariant='white'
            textSize='xl'
            textWeight={'700'}
            style={DateInputStyles().confirmButtonStyle}
          />

          {allowClear && value && (
            <Pressable
              testID='date-input-clear-button'
              onPress={handleClearPress}
              text={formatMessage('COMMON.CLEAR')}
              textVariant='white'
              textSize='xl'
              textWeight={'700'}
              style={DateInputStyles().cancelButtonStyle}
            />
          )}
        </Item>
      ),
      children: (
        <Item size='full' alignItemsCenter justifyContentCenter backgroundColor='white'>
          <DatePicker
            testID={testID ?? 'date-input-test-id'}
            mode={mode?.replace('modal-', '') as any}
            date={value ? new Date(value) : new Date()}
            minimumDate={minimumDate}
            maximumDate={maximumDate}
            theme='light'
            onDateChange={(date) => (externalDate = date)}
            {...props}
          />
        </Item>
      ),
    },
    []
  )
  const {showModal} = useModal()

  const showModalDatePicker = () =>
    showModal({
      id: ModalIdEnum.DatePickerModal,
      testID: 'date-input-modal-picker-test-id',
      visible: true,
      height: modalHeightCalculator(data?.length),
      slideToClose: true,
      closeButton: false,
      onClose: () => hideModal(ModalIdEnum.DatePickerModal),
      headerBackgroundColor: 'grey-900',
      panResponderBackgroundColor: 'grey-900',
      backgroundColor: colors['grey-900'],
      header: {
        height: 40,
        leftIcon: {
          testID: 'cancel-button-test-id',
          name: 'CHEVRON_LEFT',
          mode: 'stroke',
          width: 24,
          height: 24,
          strokeWidth: 2,
          backgroundColor: colors['grey-800'],
          borderColor: colors.primary,
          borderRadius: 24,
          alignItemsCenter: true,
          onPress: () => hideModal(ModalIdEnum.DatePickerModal),
        },
        title: formatMessage('COMMON.TRANSACTION.DATE'),
      },
      children: (
        <Item paddingTop={16} flex={1} backgroundColor={colors['grey-900']}>
          <Separator width='full' height={1} backgroundColor={colors['neutral-grey-alpha-200']} />
          <List
            height={setHeight(modalHeightCalculator(data?.length))}
            data={
              data
                ? [
                    ...data,
                    {
                      hideRightIcon: true,
                      label: formatMessage('COMMON.SELECT.DATE'),
                      action: () => {
                        hideModal(ModalIdEnum.DatePickerModal)
                        showNativeDatePicker()
                      },
                    },
                  ]
                : []
            }
            renderItem={({item}) => {
              const selected = isSelected(item)

              return (
                <>
                  <Pressable
                    testID='date-input-modal-picker-item-test-id'
                    onPress={() => {
                      if (item.value) {
                        externalDate = item.value
                        handleConfirmPress(item)
                        Modal.hideModal()
                      } else item.action()
                    }}
                    height={52}
                    justifyContentCenter
                    paddingHorizontal={24}>
                    <Item row size='full' relative alignItemsCenter>
                      <Label variant='white' fontSize='lg' fontFamily='Markpro-Medium'>
                        {item.label}
                      </Label>

                      {!item.hideRightIcon && (
                        <Item
                          width={24}
                          height={24}
                          borderRadius={24}
                          borderWidth={2}
                          borderColor={selected ? colors.primary : colors['grey-200']}
                          justifyContentCenter
                          alignItemsCenter
                          absolute
                          right={0}>
                          {selected && (
                            <Item
                              width={16}
                              height={16}
                              borderRadius={16}
                              backgroundColor={colors.primary}
                              // eslint-disable-next-line react-native/no-inline-styles
                              style={{
                                shadowColor: colors['neutral-blue-soft'],
                                shadowOffset: {
                                  width: 0,
                                  height: 0,
                                },
                                shadowOpacity: 1,
                                shadowRadius: 1,
                                borderWidth: 2,
                                borderColor: colors['neutral-blue-soft'],
                              }}
                            />
                          )}
                        </Item>
                      )}
                    </Item>
                  </Pressable>
                  <Separator
                    width='full'
                    height={1}
                    backgroundColor={colors['neutral-grey-alpha-200']}
                  />
                </>
              )
            }}
          />
        </Item>
      ),
    })

  useEffect(() => {
    externalDate = new Date(value ? new Date(value) : new Date())

    if (value) handleConfirmPress(externalDate)

    return () => {
      externalDate = new Date()
    }
  }, [])

  return (
    <>
      {viewType === 'input' && (
        <Item row>
          <Pressable
            size='full'
            height={useInputStyleNormalizer({size}).height}
            testID={testID ?? 'date-input-value-container'}
            flex={0}
            justifyContentCenter
            textSize='lg'
            text={
              data && value
                ? (data.find((item) => {
                    return isSelected(item)
                  })?.label ??
                  formatByDateMode(
                    mode.replace('modal-', '') as 'date' | 'time' | 'datetime',
                    value
                  ))
                : value
                  ? formatByDateMode(mode, value)
                  : placeholder
            }
            textStyle={DateInputStyles().pressableTextStyle}
            style={StyleSheet.flatten([defaultTextInputStyle])}
            textVariant={theme === 'dark' ? 'white' : value ? 'grey-900' : 'grey-200'}
            onPress={() => {
              if (disabled) return

              mode.includes('modal') ? showModalDatePicker() : showNativeDatePicker()
            }}
          />

          <Pressable
            onPress={() => {
              if (disabled) return

              mode.includes('modal') ? showModalDatePicker() : showNativeDatePicker()
            }}>
            <Item absolute right={0} bottom={useInputStyleNormalizer({size}).icon.paddingBottom}>
              <Icon
                name='CALENDAR'
                width={useInputStyleNormalizer({size}).icon.width}
                height={useInputStyleNormalizer({size}).icon.height}
                color={colors['grey-200']}
                mode='stroke'
                strokeWidth={2}
              />
            </Item>
          </Pressable>
        </Item>
      )}

      {viewType === 'button' && (
        <Button
          testID={testID ?? 'date-input-test-id'}
          size={{height: props?.height ?? 'md'}}
          leftIcon={{
            name: props.icon?.name ?? 'CALENDAR',
            width: 20,
            height: 20,
            strokeWidth: 1.5,
          }}
          onPress={() => {
            !disabled && mode.includes('modal') ? showModalDatePicker() : showNativeDatePicker()
          }}>
          {data && value
            ? (data.find((item) => {
                return isSelected(item)
              })?.label ??
              formatByDateMode(mode.replace('modal-', '') as 'date' | 'time' | 'datetime', value))
            : value
              ? formatByDateMode(mode, value)
              : placeholder}
        </Button>
      )}
    </>
  )
}
