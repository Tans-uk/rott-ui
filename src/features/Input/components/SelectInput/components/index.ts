export * from './SelectInput'
export * from './SelectInputModalComponent'
