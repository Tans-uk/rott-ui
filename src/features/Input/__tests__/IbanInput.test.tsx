import {IbanInput} from '../components'

import {fireEvent, render, waitFor} from '@utils'

describe('IBAN Input -> Custom Input', () => {
  const inputTestId = 'iban-input-test-id'
  const clearIbanIconTestId = 'clear-iban-icon-test-id'
  const ibanIconTestId = 'iban-icon-test-id'

  it('ilk render anında snapshot ile eşleşmeli', () => {
    const renderedInput = render(<IbanInput name='test' testID={inputTestId} />)

    expect(renderedInput).toMatchSnapshot()
  })

  it('ilk renderlandiginda ilk renderlandiginda içerik boş olmalı', () => {
    const {getByTestId} = render(<IbanInput name='test' testID={inputTestId} />)

    const inputElement = getByTestId(inputTestId)

    expect(inputElement).toHaveProp('value', '')
  })

  it('ilk renderlandiginda icerik temizleme iconu gorunmemeli', () => {
    const {queryByTestId} = render(<IbanInput name='test' testID={inputTestId} value='' />)

    const clearInputElement = queryByTestId(clearIbanIconTestId)

    expect(clearInputElement).toBeNull()
  })

  // TODO: Varsayilan TR texti yazildigindan bu test gecersiz
  it('içerik boş olduğunda temizleme iconu görünmemeli', () => {
    const onChangeTextMock = jest.fn()
    const {getByTestId, queryByTestId} = render(
      <IbanInput name='test' onChangeText={onChangeTextMock} value='TR' />
    )

    const ibanInput = getByTestId('iban-input-test-id')
    expect(ibanInput).toBeTruthy()

    const clearIcon = queryByTestId('clear-iban-icon-test-id')
    expect(clearIcon).toBeNull()

    fireEvent.changeText(ibanInput, '')

    expect(queryByTestId('clear-iban-icon-test-id')).toBeNull()
  })

  it('içerik boş olduğunda temizleme iconu görünmeli', async () => {
    const onChangeTextMock = jest.fn()
    const {getByTestId, queryByTestId, rerender} = render(
      <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
    )

    let inputElement = getByTestId(inputTestId)
    await waitFor(() => {
      fireEvent.changeText(inputElement, 'TR123')
    })

    let clearInputElement = getByTestId(clearIbanIconTestId)
    let iconElement = getByTestId(ibanIconTestId)

    expect(clearInputElement).toBeTruthy()
    expect(iconElement).toHaveProp('name', 'REMOVE_CIRCLE')

    expect(onChangeTextMock).toHaveBeenCalledWith('TR123')

    await waitFor(() => {
      inputElement = getByTestId(inputTestId)

      fireEvent.changeText(inputElement, '')
    })

    expect(onChangeTextMock).toHaveBeenCalledWith('TR')

    rerender(
      <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} value='TR' />
    )
    let clearInputElement2 = queryByTestId(clearIbanIconTestId)

    expect(clearInputElement2).toBeNull()
  })

  it('TR digerlerinden sonraki degerler sadece numeric karakterleri kabul etmeli', async () => {
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(
      <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
    )

    await waitFor(() => {
      const inputElement = getByTestId(inputTestId)
      fireEvent.changeText(inputElement, 'A*a123 334--34')
    })

    expect(onChangeTextMock).toHaveBeenCalledWith('TR12333434')
  })

  describe('kopyalanan IBAN yapıştırıldığında', () => {
    it('Bosluklar trimlenmeli', async () => {
      const onChangeTextMock = jest.fn()
      const {getByTestId} = render(
        <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
      )

      await waitFor(() => {
        const inputElement = getByTestId(inputTestId)
        fireEvent.changeText(inputElement, 'aaA*a   123')
      })

      expect(onChangeTextMock).toHaveBeenCalledWith('TR123')
    })
    it('Sadece Sayi gelmisse basina TR getirerek dogru format saglanmali', async () => {
      const onChangeTextMock = jest.fn()
      const {getByTestId} = render(
        <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
      )

      await waitFor(() => {
        const inputElement = getByTestId(inputTestId)
        fireEvent.changeText(inputElement, '123')
      })

      expect(onChangeTextMock).toHaveBeenCalledWith('TR123')
    })
    it('yapistirilan degerde TR ile gelmisse dogru format saglanmali', async () => {
      const onChangeTextMock = jest.fn()
      const {getByTestId} = render(
        <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
      )

      await waitFor(() => {
        const inputElement = getByTestId(inputTestId)
        fireEvent.changeText(inputElement, 'TR99 8888 7777 6666 5555 4444 33')
      })

      expect(onChangeTextMock).toHaveBeenCalledWith('TR998888777766665555444433')
    })

    it('yapistirilan degerde TR yok ise dogru format saglanmali', async () => {
      const onChangeTextMock = jest.fn()
      const {getByTestId} = render(
        <IbanInput name='test' testID={inputTestId} onChangeText={onChangeTextMock} />
      )

      await waitFor(() => {
        const inputElement = getByTestId(inputTestId)
        fireEvent.changeText(inputElement, '99 8888 7777 6666 5555 4444 33')
      })

      expect(onChangeTextMock).toHaveBeenCalledWith('TR998888777766665555444433')
    })
  })

  it('Iban input render olduğu zaman klavye olarak number-pad ekranda görülmeli', () => {
    const {getByTestId} = render(<IbanInput name='test' testID={inputTestId} />)
    const inputElement = getByTestId(inputTestId)

    expect(inputElement).toHaveProp('keyboardType', 'number-pad')
  })
})
