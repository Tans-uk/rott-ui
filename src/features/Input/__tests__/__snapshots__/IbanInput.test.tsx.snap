// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IBAN Input -> Custom Input ilk render anında snapshot ile eşleşmeli 1`] = `
<View
  flex={1}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    {
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <KeyboardProvider
    enabled={true}
  >
    <View
      flex={1}
      pointerEvents="box-none"
      style={
        {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <View
          style={
            {
              "flex": undefined,
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            editable={true}
            keyboardType="number-pad"
            maxLength={34}
            name="test"
            numberOfLines={1}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="TR00 0000 0000 0000 0000 0000 00"
            style={
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontFamily": "Markpro-Medium",
                "fontSize": 14,
                "height": 89,
                "letterSpacing": 0.5,
                "width": "100%",
              }
            }
            testID="iban-input-test-id"
            value=""
          />
          <View
            absolute={true}
            bottom={16}
            right={0}
            style={
              {
                "bottom": 31,
                "flex": undefined,
                "flexDirection": "column",
                "position": "absolute",
                "right": 0,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              animated={false}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "flex": 1,
                  "flexDirection": "column",
                }
              }
              testID="clear-iban-icon-test-id"
            >
              <View
                style={
                  {
                    "flex": undefined,
                    "flexDirection": "column",
                  }
                }
              >
                <SvgMock
                  fill="rgba(161, 173, 175, 1)"
                  height={46}
                  stroke="rgba(161, 173, 175, 1)"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={0}
                  width={46}
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "flexShrink": 1,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "flexDirection": "column-reverse",
              "justifyContent": "flex-start",
              "top": 30,
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          [
            {
              "alignSelf": "center",
              "elevation": 999999,
              "flex": 0,
              "maxWidth": "100%",
              "position": "absolute",
              "zIndex": 999999,
            },
            {
              "bottom": 40,
              "flexDirection": "column",
              "justifyContent": "flex-end",
              "width": 750,
            },
          ]
        }
      >
        <View
          style={
            {
              "bottom": 0,
            }
          }
        >
          <RCTSafeAreaView />
        </View>
      </View>
    </View>
  </KeyboardProvider>
</View>
`;
